### PR TITLE
Fix a race condition in getting the trunk name via its mac

### PR DESCRIPTION
**Pull request by fenxiong**
During setting up a branch interface, the branch can have the same mac address as the trunk for a short period of time, and we may end up picking the branch's name when trying to get the trunk name by mac. Fixing it by picking the shortest name available since branch name is strictly longer than trunk.

Unit test added and passed. The e2e tests via vpc-branch-eni-e2e-tests target passed. Manual test in progress.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: network/eni/eni.go [hunk 1]
```diff
@@ -102,6 +97,27 @@ func (eni *ENI) AttachToLink() error {
 	return nil
 }
 
+// getInterfaceByMac gets the interface with specified mac address, and picks the one with shortest name
+// if there are multiple matches, which can happen temporarily when setting up a branch ENI, in which it
+// can have the same mac address as the trunk during branch.AttachToLink.
+func getInterfaceByMac(macAddress net.HardwareAddr, interfaces []net.Interface) *net.Interface {
```
Comment (ofiliz) [inline network/eni/eni.go]: Can you please call this getInterfaceByMACAddress, to comply with the Go naming conventions? Also please move to the end of the file to keep the ENI class methods together.

File: network/eni/eni.go [hunk 2]
```diff
@@ -102,6 +97,27 @@ func (eni *ENI) AttachToLink() error {
 	return nil
 }
 
+// getInterfaceByMac gets the interface with specified mac address, and picks the one with shortest name
```
Comment (ofiliz) [inline network/eni/eni.go]: Let's not discuss the trunk vs. branch MAC address here. The ENI class doesn't know about trunks etc.
// getInterfaceByMACAddress returns the interface with the specified MAC address.

Move the rest of your comment block to just before the for loop (line 106):
// If there are multiple matches, pick the one with the shortest name.

File: network/eni/eni.go [hunk 3]
```diff
@@ -102,6 +97,27 @@ func (eni *ENI) AttachToLink() error {
...
+// if there are multiple matches, which can happen temporarily when setting up a branch ENI, in which it
+// can have the same mac address as the trunk during branch.AttachToLink.
+func getInterfaceByMac(macAddress net.HardwareAddr, interfaces []net.Interface) *net.Interface {
+	var chosenInterface net.Interface
```
Comment (ofiliz) [inline network/eni/eni.go]: If you define this as a pointer (var chosen *net.Interface), then

the if check at line 108 can become simply
if chosen == nil || len(chosen.Name) > len(i.Name)

line 109 becomes a pointer copy instead of a struct copy

you can remove the if block lines 114-116.

File: network/eni/eni.go [hunk 4]
```diff
@@ -82,12 +82,7 @@ func (eni *ENI) AttachToLink() error {
 			return err
 		}
 
-		for _, i := range interfaces {
-			if vpc.CompareMACAddress(i.HardwareAddr, eni.macAddress) {
-				iface = &i
-				break
-			}
-		}
+		iface = getInterfaceByMac(eni.macAddress, interfaces)
```
Comment (ofiliz) [inline network/eni/eni.go]: Should we move the net.Interfaces() call to the new getInterfaceByMACAddress function? That way we wouldn't have to pass it as an argument.

File: network/eni/eni_test.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+		{
+			Index: 2,
+			Name: "eth1",
+			HardwareAddr: mac,
+		},
+		{
+			Index: 3,
+			Name: "eth1.1.1",
+			HardwareAddr: mac,
+		},
```
Comment (ofiliz) [inline network/eni/eni_test.go]: Thanks for adding tests. For testing an interesting corner case, you can add one more interface here with a three letter name but a different MAC address. To make sure MAC address match is prioritized.

File: network/eni/eni_test.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package eni
```
Comment (ofiliz) [inline network/eni/eni_test.go]: Should we add the copyright header to the top of the file?

File: network/eni/eni.go (hunk 1 above)
Comment (fenxiong) [reply network/eni/eni.go]: updated

File: network/eni/eni.go (hunk 2 above)
Comment (fenxiong) [reply network/eni/eni.go]: updated

File: network/eni/eni.go (hunk 3 above)
Comment (fenxiong) [reply network/eni/eni.go]: i originally did that. then i found the variable "i" in the for loop shares the same address, so i can't just save the address.

File: network/eni/eni.go (hunk 4 above)
Comment (fenxiong) [reply network/eni/eni.go]: that makes it hard to add unit test i think..

File: network/eni/eni_test.go (hunk 5 above)
Comment (fenxiong) [reply network/eni/eni_test.go]: updated the test

File: network/eni/eni_test.go (hunk 6 above)
Comment (fenxiong) [reply network/eni/eni_test.go]: good catch, added

File: network/eni/eni.go (hunk 3 above)
Comment (fenxiong) [reply network/eni/eni.go]: discussed offline. fixed.
